sale_certificat_typology/models/sale_order.py

```diff
@@ -0,0 +1,150 @@
+                    "Warning: in this order, some products require certificates !"
+                )
+            sale.sale_warn_msg = sale_warn_msg
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    certificat_item_ids = fields.Many2many(
+        comodel_name="sale.order.certificat.item", string="Linked certificat items"
+    )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        lines = super().create(vals_list)
+        for line, vals in zip(lines, vals_list):
+            if "order_id" in vals and "product_id" in vals:
+                self._add_certificat_items(line, vals.get("product_id"))
+        return lines
+
+    def write(self, values):
+        for line in self:
+            if "product_id" in values:
+                # remove certificat item for old product
+                self._remove_certificat_items(line)
+                # add certificat item for new product
+                self._add_certificat_items(line, values.get("product_id"))
+        return super().write(values)
+
+    def unlink(self):
+        for line in self:
+            self._remove_certificat_items(line)
+        return super().unlink()
+
+    def _add_certificat_items(self, line, product_id):
+        product = self.env["product.product"].browse(product_id)
+        for certificat in product.required_certificat_ids:
+            certificat_item = self.env["sale.order.certificat.item"].search(
+                [
+                    ("order_id", "=", line.order_id.id),
+                    ("certificat_typology_id", "=", certificat.id),
+                ],
+                limit=1,
+            )
+            if certificat_item:
+                certificat_item.write({"order_line_ids": [(4, line.id, 0)]})
+            else:
+                self.env["sale.order.certificat.item"].create(
+                    {
+                        "order_id": line.order_id.id,
+                        "order_line_ids": [(6, 0, line.ids)],
+                        "certificat_typology_id": certificat.id,
+                        "company_id": line.company_id.id,
+                    }
+                )
+
+    def _remove_certificat_items(self, line):
+        certificat_items = self.env["sale.order.certificat.item"].search(
+            [("order_id", "=", line.order_id.id), ("order_line_ids", "in", line.id)]
+        )
+        for item in certificat_items:
+            if len(item.order_line_ids) == 1:
+                item.unlink()
+            else:
+                item.write({"order_line_ids": [(3, line.id, 0)]})
+
+
+class SaleOrderCertificatItem(models.Model):
+    _name = "sale.order.certificat.item"
+    _description = "Certificat Items in Sale Order"
+
+    order_id = fields.Many2one(
+        comodel_name="sale.order", string="Linked order", required=True
+    )
+    order_line_ids = fields.Many2many(
+        comodel_name="sale.order.line", string="Linked order lines", required=True
+    )
```

Comment:
not needed

Reply:
@sebastienbeau , I made the suggested changes.